File: app/src/main/java/org/mifos/mobilebanking/presenters/LoginPresenter.java
package org.mifos.mobilebanking.presenters;

import android.content.Context;
import android.content.res.Resources;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.api.BaseApiManager;
import org.mifos.mobilebanking.api.DataManager;
import org.mifos.mobilebanking.api.local.PreferencesHelper;
import org.mifos.mobilebanking.injection.ApplicationContext;
import org.mifos.mobilebanking.models.Page;
import org.mifos.mobilebanking.models.User;
import org.mifos.mobilebanking.models.client.Client;
import org.mifos.mobilebanking.presenters.base.BasePresenter;
import org.mifos.mobilebanking.ui.views.LoginView;
import org.mifos.mobilebanking.utils.Constants;
import org.mifos.mobilebanking.utils.MFErrorParser;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.plugins.RxJavaPlugins;
import io.reactivex.schedulers.Schedulers;
import retrofit2.HttpException;

/**
 * @author Vishwajeet
 * @since 05/06/16
 */
public class LoginPresenter extends BasePresenter<LoginView> {

    private final DataManager dataManager;
    private PreferencesHelper preferencesHelper;
    private CompositeDisposable compositeDisposable;

    /**
     * Initialises the LoginPresenter by automatically injecting an instance of
     * {@link DataManager} and {@link Context}.
     *
     * @param dataManager DataManager class that provides access to the data
     *                    via the API.
     * @param context     Context of the view attached to the presenter. In this case
     *                    it is that of an {@link android.support.v7.app.AppCompatActivity}
     */
    @Inject
    public LoginPresenter(DataManager dataManager, @ApplicationContext Context context) {
        super(context);
        this.dataManager = dataManager;
        preferencesHelper = this.dataManager.getPreferencesHelper();
        compositeDisposable = new CompositeDisposable();
    }

    @Override
    public void attachView(LoginView mvpView) {
        super.attachView(mvpView);
    }

    @Override
    public void detachView() {
        super.detachView();
        compositeDisposable.clear();
    }

    /**
     * This method validates the username and password entered by the user
     * and reports any errors that might exists in any of the inputs.
     * If there are no errors, then we attempt to authenticate the user from
     * the server and then persist the authentication data if we successfully
     * authenticate the credentials or notify the view about any errors.
     *
     * @param username Username of the user trying to login.
     * @param password Password of the user trying to login.
     */
    public void login(final String username, final String password) {
        checkViewAttached();
        if (isCredentialsValid(username, password)) {
            getMvpView().showProgress();
            compositeDisposable.add(dataManager.login(username, password)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeOn(Schedulers.io())
                    .subscribeWith(new DisposableObserver<User>() {
                        @Override
                        public void onComplete() {

                        }

                        @Override
                        public void onError(Throwable e) {
                            getMvpView().hideProgress();
                            String errorMessage;
                            try {
                                if (e instanceof HttpException) {
                                    if (((HttpException) e).code() == 503) {
                                        getMvpView().
                                                showMessage(context.
                                                        getString(R.string.error_server_down));
                                    } else {
                                        errorMessage =
                                                ((HttpException) e).response().errorBody().string();
                                        getMvpView()
                                                .showMessage(MFErrorParser.parseError(errorMessage)
                                                .getDeveloperMessage());
                                    }
                                }
                            } catch (Throwable throwable) {
                                RxJavaPlugins.getErrorHandler();
                            }
                        }

                        @Override
                        public void onNext(User user) {
                            if (user != null) {
                                final String userName = user.getUsername();
                                final long userID = user.getUserId();
                                final String authToken = Constants.BASIC +
                                        user.getBase64EncodedAuthenticationKey();
                                saveAuthenticationTokenForSession(userName, userID, authToken);
                                getMvpView().onLoginSuccess(userName);
                            } else {
                                getMvpView().hideProgress();
                            }
                        }
                    })
            );
        }
    }

    /**
     * This method fetching the Client, associated with current Access Token.
     */
    public void loadClient() {
        checkViewAttached();
        compositeDisposable.add(dataManager.getClients()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DisposableObserver<Page<Client>>() {
                    @Override
                    public void onComplete() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().hideProgress();
                        if (((HttpException) e).code() == 401) {
                            getMvpView().showMessage(context.getString(R.string.
                                    unauthorized_client));
                        } else {
                            getMvpView().showMessage(context.getString(R.string.
                                    error_fetching_client));
                        }
                        preferencesHelper.clear();
                        reInitializeService();
                    }

                    @Override
                    public void onNext(Page<Client> clientPage) {
                        getMvpView().hideProgress();
                        if (clientPage.getPageItems().size() != 0) {
                            long clientId = clientPage.getPageItems().get(0).getId();
                            preferencesHelper.setClientId(clientId);
                            dataManager.setClientId(clientId);
                            reInitializeService();
                            getMvpView().showPassCodeActivity();
                        } else {
                            getMvpView().showMessage(context
                                    .getString(R.string.error_client_not_found));
                        }
                    }
                })
        );
    }


    private boolean isCredentialsValid(final String username, final String password) {
        boolean credentialValid = true;
        final Resources resources = context.getResources();
        final String correctUsername = username.replaceFirst("\\s++$", "").trim();
        if (username == null || username.matches("\\s*") || username.isEmpty()) {
            getMvpView().showUsernameError(context.getString(R.string.error_validation_blank,
                    context.getString(R.string.username)));
            credentialValid = false;
        } else if (username.length() < 5) {
            getMvpView().showUsernameError(context.getString(R.string.error_validation_minimum_chars
                    , resources.getString(R.string.username), resources.getInteger(R.integer.
                            username_minimum_length)));
            credentialValid = false;
        } else if (correctUsername.contains(" ")) {
            getMvpView().showUsernameError(context.getString(
                    R.string.error_validation_cannot_contain_spaces,
                    resources.getString(R.string.username),
                    context.getString(R.string.not_contain_username)));
            credentialValid = false;
        } else {
            getMvpView().clearUsernameError();
        }

        if (password == null || password.isEmpty()) {
            getMvpView().showPasswordError(context.getString(R.string.error_validation_blank,
                    context.getString(R.string.password)));
            credentialValid = false;
        } else if (password.length() < 6) {
            getMvpView().showPasswordError(context.getString(R.string.error_validation_minimum_chars
                    , resources.getString(R.string.password), resources.getInteger(R.integer.
                            password_minimum_length)));
            credentialValid = false;
        } else {
            getMvpView().clearPasswordError();
        }

        return credentialValid;
    }

    /**
     * Save the authentication token from the server and the user ID.
     * The authentication token would be used for accessing the authenticated
     * APIs.
     *
     * @param userID    - The userID of the user to be saved.
     * @param authToken - The authentication token to be saved.
     */
    private void saveAuthenticationTokenForSession(String userName, long userID, String authToken) {
        preferencesHelper.setUserName(userName);
        preferencesHelper.setUserId(userID);
        preferencesHelper.saveToken(authToken);
        reInitializeService();
    }

    private void reInitializeService() {
        BaseApiManager.createService(preferencesHelper.getBaseUrl(), preferencesHelper.getTenant(),
                preferencesHelper.getToken());
    }
}

File: app/src/main/java/org/mifos/mobile/presenters/LoginPresenter.kt
package org.mifos.mobile.presenters

import android.content.Context

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableObserver
import io.reactivex.plugins.RxJavaPlugins
import io.reactivex.schedulers.Schedulers

import org.mifos.mobile.R
import org.mifos.mobile.api.BaseApiManager
import org.mifos.mobile.api.DataManager
import org.mifos.mobile.api.local.PreferencesHelper
import org.mifos.mobile.injection.ApplicationContext
import org.mifos.mobile.models.Page
import org.mifos.mobile.models.User
import org.mifos.mobile.models.client.Client
import org.mifos.mobile.models.payload.LoginPayload
import org.mifos.mobile.presenters.base.BasePresenter
import org.mifos.mobile.ui.views.LoginView
import org.mifos.mobile.utils.Constants
import org.mifos.mobile.utils.MFErrorParser

import retrofit2.HttpException

import javax.inject.Inject

/**
 * @author Vishwajeet
 * @since 05/06/16
 */
class LoginPresenter @Inject constructor(private val dataManager: DataManager, @ApplicationContext context: Context?) :
        BasePresenter<LoginView?>(context) {

    private val preferencesHelper: PreferencesHelper = dataManager.preferencesHelper
    private val compositeDisposable: CompositeDisposable = CompositeDisposable()
    override fun attachView(mvpView: LoginView?) {
        super.attachView(mvpView)
    }

    override fun detachView() {
        super.detachView()
        compositeDisposable.clear()
    }

    /**
     * This method validates the username and password entered by the user
     * and reports any errors that might exists in any of the inputs.
     * If there are no errors, then we attempt to authenticate the user from
     * the server and then persist the authentication data if we successfully
     * authenticate the credentials or notify the view about any errors.
     */
    fun login(loginPayload: LoginPayload?) {
        checkViewAttached()
        if (isCredentialsValid(loginPayload)) {
            mvpView?.showProgress()
            compositeDisposable.add(dataManager.login(loginPayload)
                    ?.observeOn(AndroidSchedulers.mainThread())
                    ?.subscribeOn(Schedulers.io())!!
                    .subscribeWith(object : DisposableObserver<User?>() {
                        override fun onComplete() {}
                        override fun onError(e: Throwable) {
                            mvpView?.hideProgress()
                            val errorMessage: String
                            try {
                                if (e is HttpException) {
                                    if (e.code() == 503) {
                                        mvpView?.showMessage(context?.getString(R.string.error_server_down))
                                    } else {
                                        errorMessage = e.response().errorBody().string()
                                        mvpView
                                                ?.showMessage(MFErrorParser.parseError(errorMessage)
                                                        .developerMessage)
                                    }
                                }
                            } catch (throwable: Throwable) {
                                RxJavaPlugins.getErrorHandler()
                            }
                        }

                        override fun onNext(user: User) {
                            val userName = user.username
                            val userID = user.userId
                            val authToken = Constants.BASIC +
                                    user.base64EncodedAuthenticationKey
                            saveAuthenticationTokenForSession(userName, userID, authToken)
                            mvpView?.onLoginSuccess(userName)
                        }
                    })
            )
        }
    }

    /**
     * This method fetching the Client, associated with current Access Token.
     */
    fun loadClient() {
        checkViewAttached()
        dataManager.clients
                ?.observeOn(AndroidSchedulers.mainThread())
                ?.subscribeOn(Schedulers.io())
                ?.subscribeWith(object : DisposableObserver<Page<Client?>?>() {
                    override fun onComplete() {}
                    override fun onError(e: Throwable) {
                        mvpView?.hideProgress()
                        if ((e as HttpException).code() == 401) {
                            mvpView?.showMessage(context?.getString(R.string.unauthorized_client))
                        } else {
                            mvpView?.showMessage(context?.getString(R.string.error_fetching_client))
                        }
                        preferencesHelper.clear()
                        reInitializeService()
                    }

                    override fun onNext(clientPage: Page<Client?>) {
                        mvpView?.hideProgress()
                        if (clientPage.pageItems.isNotEmpty()) {
                            val clientId = clientPage.pageItems[0]?.id?.toLong()
                            preferencesHelper.clientId = clientId
                            dataManager.clientId = clientId
                            reInitializeService()
                            mvpView?.showPassCodeActivity()
                        } else {
                            mvpView?.showMessage(context
                                    ?.getString(R.string.error_client_not_found))
                        }
                    }
                })?.let {
                    compositeDisposable.add(it
                    )
                }
    }

    private fun isCredentialsValid(loginPayload: LoginPayload?): Boolean {
        val username: String = loginPayload?.username.toString()
        val password: String = loginPayload?.password.toString()
        var credentialValid = true
        val resources = context?.resources
        val correctUsername = username.replaceFirst("\\s++$".toRegex(), "").trim { it <= ' ' }
        when {
            username.isEmpty() -> {
                mvpView?.showUsernameError(context?.getString(R.string.error_validation_blank,
                        context?.getString(R.string.username)))
                credentialValid = false
            }
            username.length < 5 -> {
                mvpView?.showUsernameError(context?.getString(R.string.error_validation_minimum_chars, resources?.getString(R.string.username), resources?.getInteger(R.integer.username_minimum_length)))
                credentialValid = false
            }
            correctUsername.contains(" ") -> {
                mvpView?.showUsernameError(context?.getString(
                        R.string.error_validation_cannot_contain_spaces,
                        resources?.getString(R.string.username),
                        context?.getString(R.string.not_contain_username)))
                credentialValid = false
            }
            else -> {
                mvpView?.clearUsernameError()
            }
        }
        when {
            password.isEmpty() -> {
                mvpView?.showPasswordError(context?.getString(R.string.error_validation_blank,
                        context?.getString(R.string.password)))
                credentialValid = false
            }
            password.length < 6 -> {
                mvpView?.showPasswordError(context?.getString(R.string.error_validation_minimum_chars,
                        resources?.getString(R.string.password),
                        resources?.getInteger(R.integer.password_minimum_length)))
                credentialValid = false
            }
            else -> {
                mvpView?.clearPasswordError()
            }
        }
        return credentialValid
    }

    /**
     * Save the authentication token from the server and the user ID.
     * The authentication token would be used for accessing the authenticated
     * APIs.
     *
     * @param userID    - The userID of the user to be saved.
     * @param authToken - The authentication token to be saved.
     */
    private fun saveAuthenticationTokenForSession(userName: String?, userID: Long, authToken: String) {
        preferencesHelper.userName = userName
        preferencesHelper.userId = userID
        preferencesHelper.saveToken(authToken)
        reInitializeService()
    }

    private fun reInitializeService() {
        BaseApiManager.createService(preferencesHelper.baseUrl, preferencesHelper.tenant,
                preferencesHelper.token)
    }

}